### ISSUE TITLE
Missing contacts in Fossify Phone contact lists

### ISSUE THREAD

**Issue by igor-cali** [FossifyOrg/Phone#73]
### Checklist

- [X] I can reproduce the bug with the latest version given [here](https://github.com/FossifyOrg/Phone/releases/latest).
- [X] I made sure that there are **no existing issues** - [open](https://github.com/FossifyOrg/Phone/issues) or [closed](https://github.com/FossifyOrg/Phone/issues?q=is%3Aissue+is%3Aclosed) - to which I could contribute my information.
- [X] I made sure that there are **no existing discussions** - [open](https://github.com/FossifyOrg/Phone/discussions) or [closed](https://github.com/FossifyOrg/Phone/discussions?discussions_q=is%3Aclosed) - to which I could contribute my information.
- [X] I have read the FAQs inside the app (Menu -> About -> FAQs) and my problem isn't listed.
- [X] **I have taken the time to fill in all the required details. I understand that the bug report will be dismissed otherwise.**
- [X] This issue contains only one bug.
- [X] I have read and understood the [contribution guidelines](https://github.com/FossifyOrg/Phone/blob/master/CONTRIBUTING.md).

### Affected app version

1.0.0

### Affected Android/Custom ROM version

Android 13

### Affected device model

Fairphone 3

### How did you install the app?

F-Droid / IzzyOnDroid

### Steps to reproduce the bug

- Open app (`Contacts` or `History` tab is shown)
- Click the three dots menu of a Contact included in more than one account (e.g. _Contact list / CardDAV_, _Telegram_ and/or _Whatsapp_)
- Select menu item `Show contact details`

Check displayed information

### Expected behavior

All contact information are shown in the contact details (from the `Contacts` or `History` tabs), as it happens with the `Fossify Contacts` app.

### Actual behavior

As an example, for a contact of my I find three different entries in the `Fossify Contacts` app:
    one for the contact from _Nextcloud_ (_CardDAV_ account from `DAV5x`);
    one for the _Telegram_ contact;
    one for the Whatsapp contact.

For the same person, in the Phone app I just have one entry, related to _Telegram_.

### Additional information

My contacts are in my private Nextcloud instance. Using **DAVx5** Android App to sync them.
This issue is the same as an [original Simple Dialer issue](https://github.com/SimpleMobileTools/Simple-Dialer/issues/230#).

**Comment by Aga-C** [comment]
Do I understand correctly, that in Contacts app you have turned off merging contacts? And you'd like to have opening contact details in Phone work like this, that if you're filtering by WhatsApp contact list, then it should open details of contact from WhatsApp, not the default one that Contacts app open? 

It works this way, because in Phone all contacts are merged, so even if you filter by phone storage, the app sends an intent to Contact app to open the contact without telling from which storage it should be. And, if in Contacts app you have turned off merging, it just opens one of the instances of the contact.

**Comment by igor-cali** [comment]
> Do I understand correctly, that in Contacts app you have turned off merging contacts?

@Aga-C No, in my Contacts app the `Merge duplicate contacts` checkbox is checked.
Following your question, I tried to uncheck it. The behaviour does not change (the issue still occurs).

**Comment by Piprr** [comment]
Even with backups, I have to manually re enter contacts again and again.

**Comment by chesio** [comment]
> It works this way, because in Phone all contacts are merged, so even if you filter by phone storage, the app sends an intent to Contact app to open the contact without telling from which storage it should be. And, if in Contacts app you have turned off merging, it just opens one of the instances of the contact.

@Aga-C isn't there another way to implement it? It's a mess right now as soon as one starts editing contacts from within Phone app and have multiple contact sources because of apps like WhatsApp, Signal etc.

I have all my daily contacts in CardDAV storage, but some are also present in WhatsApp and Signal. I don't need to access my WhatsApp and Signal contacts via Contacts or Phone app, so I have only CardDAV storage active via storage filter in both apps.

A friend recently got a new number and called me. After the call I immediately added her number to my contacts - I assumed I was working with CardDAV storage (because of the active filter). Today she called again and the number was shown as unknown. I was surprised, but immediately after the call I proceeded to add her new number to existing CardDAV contact again only to find the number is already there..!

I was confused, so I went to Contacts app to check her CardDAV contact again: no, only old number was there. Only then I got the idea that it must somehow be related to contact source, so I activated display of source in contact details in both apps and: **If I open her contact from Contacts tab in Phone app, I see the WhatsApp contact (not CardDAV as I would expect). And when I attempt to add her new number to existing contact from Call History tab, I get to edit the Signal contact (not CardDAV as I would expect)**. So there is really no consistency whatsoever...

Also, the list of storage filters seems to be affected by this issue: in Contacts app the list correctly reports 304 contacts in my CardDAV storage, while in Phone app the same list reports only 279 contacts in my CardDAV storage.

**Comment by Aga-C** [comment]
@chesio 

> @Aga-C isn't there another way to implement it?

Probably something can be done, but I haven't investigated it thoroughly. Here's the code for it: https://github.com/FossifyOrg/Phone/blob/f49876ebefeee44ef2ef9bd3018cce44b8d8f622/app/src/main/kotlin/org/fossify/phone/extensions/Activity.kt#L67. There's a chance that the lookup key is wrongly fetched.

> in Contacts app the list correctly reports 304 contacts in my CardDAV storage, while in Phone app the same list reports only 279 contacts in my CardDAV storage.

Maybe these contacts don't have a phone number? In Phone, we display only contacts that have it.

**Comment by chesio** [comment]
> Probably something can be done, but I haven't investigated it thoroughly. Here's the code for it:

I'm afraid I cannot help here - I'm a developer myself, but have no experience with Android development.

> Maybe these contacts don't have a phone number? In Phone, we display only contacts that have it.

Good call, this is indeed the reason. I now somehow recall this has been mentioned in Issues before. Would a PR with new FAQ entry for this be accepted?

**Comment by naveensingh** [comment]
> Would a PR with new FAQ entry for this be accepted?

Makes sense to me.

**Comment by naveensingh** [comment]
I'm assuming you guys are using Fossify Contact.

> As an example, for a contact of mine I find three different entries in the Fossify Contacts app:
>
> - one for the contact from Nextcloud (CardDAV account from DAV5x);
> - one for the Telegram contact;
> - one for the Whatsapp contact.
>
> For the same person, in the Phone app I just have one entry, related to Telegram.

> in my Contacts app the Merge duplicate contacts checkbox is checked

@igor-cali Wouldn't those 3 entries be show up as one since you have the `Merge duplicate contacts` option enabled?

> If I open her contact from Contacts tab in Phone app, I see the WhatsApp contact (not CardDAV as I would expect).

@chesio In the Contacts app, do you also have the `Merge duplicate contacts` option checked?

> There's a chance that the lookup key is wrong

It is correct... which makes me question, have you guys tried other Contacts apps (Google, Samsung) with Fossify Phone? Does it work as expected with those apps?

More investigation is due but to me, this sounds more like a "bug" in the Contacts and the way contacts are being handled. The contact app should ask for storage location when you tap the edit button.

**Comment by igor-cali** [comment]
> [@igor-cali](https://github.com/igor-cali) Wouldn't those 3 entries be show up as one since you have the `Merge duplicate contacts` option enabled?
> 
> > If I open her contact from Contacts tab in Phone app, I see the WhatsApp contact (not CardDAV as I would expect).

Hi @naveensingh Sorry for the late reply.
It looks like the behaviour of the apps in my phone is somehow improved now. I'll need to deep dive to check if and where I still experience this with any contact.
Yes, I have the 'merge contacts' option enable in Fossify Contacts, but it depends on each contact if they are properly merged or kept separate. I don't know in details the merge logic, but it looks like phone numbers and names (and maybe prefixes) are considered... But, I think this is off topic for this issue. :)

**Comment by mattdale77** [comment]
I'm not sure where they best place to raise this is. There is new behaviour from the edit button and this is progress. But it still misses most of my use cases.

The edit button should edit the default address book and all linked contacts should be uodatedy unless they are specifically unlinked.

For example a slight name change could involve 5 or even 10 address book changes. Each in turn.

Could it be as simple as asking if linked entries should also be updated with a confirmation dialog? 

I want merged contracts to be treated as a unit unless they are intentionally split. Maybe should be a new request but it seems related

**Comment by naveensingh** [comment]
There's a request about it: https://github.com/FossifyOrg/Contacts/issues/22

Under the hood, contact aggregation isn't handled properly in both apps. That must be fixed before any "Update all linked contacts" option can be considered.

**Comment by akki42** [comment]
> Under the hood, contact aggregation isn't handled properly in both apps. That must be fixed before any "Update all linked contacts" option can be considered.

Here is another example of what is currently not working properly:

- Have contacts (with varying info) in different sources
- Turn off "merging duplicates" in Contacts
- Per contact, Phone shows only info from from one of the sources (missing the info from other sources)

Furthermore:
- If sources are filtered, Phone will only show contacts included in the active source(s), but for such contacts will **not** (always) show info from that source(s), but (sometimes) shows info from other, not activated sources instead.
- Which source info is shown appears to be random (having two sources active, for some duplicate entries info is shown from source 1, for others from source 2).


[A different issue, also related to contact aggregation, is discussed in https://github.com/FossifyOrg/Contacts/issues/78 ...]
